### PR TITLE
[EE] Improve Forwarding Department behaviour with Waiting queue feature

### PR DESCRIPTION
**Pull request by murtaza98**
## Issue(s)

**Old Behaviour:**
Currently when we forward a chat to a particular department and all the agents in target department are at maximum capacity, the chat either not getting transferred (on or after v3.13.0) or the user were getting an error saying **no-agents are currently online**(before v3.13.0).

**New Behaviour:**
In the new behaviour we're proposing, if a chat is forwarded and all agents in the target department are at maximum capacity, the chat will get transferred to the target department's waiting queue. This means that the chat will be removed from the current serving agent's chats and then will be moved to the target department's waiting queue. The visitor will see their current position within the queue.

Note: This feature will not affect the `Manual Selection` Omnichannel Routing method in any way.
 
## Steps to test or reproduce

1. Goto Admin -> Omnichannel -> Routing
a. Set `Omnichannel Routing Method` to either `Auto Selection` or `Load Balancing`. 
b. Turn on `Waiting Queue`.
c. Set `Max. number of simultaneous chats` to `1` (one)
2. Create 2 departments, for and add one agents to each one of them respectively eg: **TestDept-1** and **TestDept-2** with **Agent-1** and **Agent-2** respectively.
3. Start 2 Visitor conversation with each of the Department. After both the chats is started, forward the chat from one department to another. eg: Transfer from **Agent-1** from **TestDept-1** to **TestDept-2**. 
4. **OLD BEHAVIOUR:** If you're on version **v3.13.0 or above**, you should see that the chat is not transferred. If you're below **v1.13.0**, you should see an error message saying something like, **no agents available**.
5. **NEW BEHAVIOUR:** The chat should be transferred to the target department's waiting queue and will be removed from the old department. In our example above: Agent-1 will be removed from **TestDept-1** and will be put to **Test-Dept-2**. Its queue position will be **1**. Once **Agent-2** closes the existing chat, the forwarded chat should assigned to the **Agent-2**.

### CODE REVIEW

File: ee/app/livechat-enterprise/server/hooks/onAgentAssignmentFailed.ts [hunk 1]
```diff
@@ -1,25 +1,59 @@
 import { callbacks } from '../../../../../app/callbacks/server';
 import { LivechatInquiry, Subscriptions, LivechatRooms } from '../../../../../app/models/server';
 import { queueInquiry } from '../../../../../app/livechat/server/lib/QueueManager';
+import { RoutingManager } from '../../../../../app/livechat/server/lib/RoutingManager';
+import { settings } from '../../../../../app/settings/server';
+import { Livechat } from '../../../../../app/livechat/server/lib/Livechat';
 
-const handleOnAgentAssignmentFailed = async ({ inquiry, room }: { inquiry: any; room: any }): Promise<any> => {
-	if (!inquiry || !room || !room.onHold) {
+const handleOnAgentAssignmentFailed = async ({ inquiry, room, options }: { inquiry: any; room: any; options: { forwardingToDepartment?: { oldDepartmentId: string; transferData: any }; clienAction?: boolean} }): Promise<any> => {
```
Comment (renatobecker) [inline ee/app/livechat-enterprise/server/hooks/onAgentAssignmentFailed.ts:8]:
```suggestion
const handleOnAgentAssignmentFailed = async ({ inquiry, room, options }: { inquiry: any; room: any; options: { forwardingToDepartment?: { oldDepartmentId: string; transferData: any }; clientAction?: boolean} }): Promise<any> => {
```

**Review by renatobecker — CHANGES_REQUESTED** [review]
Overall, your code looks good.
But I found an issue that is not solved yet: When an agent attempts to transfer a chat from department A to department B and there's no agent available on department B, the agent gets the message `Trasnferred`, but the chat hasn't been transferred.

**Comment by murtaza98** [comment]
> Overall, your code looks good.
> But I found an issue that is not solved yet: When an agent attempts to transfer a chat from department A to department B and there's no agent available on department B, the agent gets the message `Trasnferred`, but the chat hasn't been transferred.

Hey @renatobecker, A new check I've added [here](https://github.com/RocketChat/Rocket.Chat/pull/22043/files#diff-90746cca82cf7830f3a5fed7352254dadb5ad47839396db5a9213c4a8d8bffdbR50) should solve the above issue.

**Comment by murtaza98** [comment]
@renatobecker I noticed that on the client we were not checking the transfer API response. As you are aware that transfer API aka (`livechat:transfer` method) returns a boolean specifying if the transfer was successful or not - however as you can see [here](https://github.com/RocketChat/Rocket.Chat/pull/22043/files#diff-7ea22d1f2dc14b7eb09df59074efe4d9a43c96cce7e9008e92e9b792fe13c56cR158), we weren't checking it due to which sometimes it was showing the `Transferred` success message even if the transfer was not successful. 

To solve the above issue, I had to do some refactoring and you might now notice that within the `handleOnAgentAssignmentFailed` callback [here](https://github.com/RocketChat/Rocket.Chat/pull/22043/files#diff-c144ad892c7f6847772bfb3e15028383d0b734cc283d908fe558fa048c060c9dR6), we're now returning the room object with a flag set incase we want this chat to be queued. Then within our `forwardRoomToDepartment` method we're checking this flag and putting the chat in queue if it's set. I think this is a better approach than my earlier solution as now we're reusing all [these](https://github.com/RocketChat/Rocket.Chat/pull/22043/files#diff-77f1b2de857a60fbc4d14cb61ccdb8fa82d46a5d283a82d894796ced9ed3585bR394-R402) code.

Let me know what you think about this? Thx

**Review by renatobecker — APPROVED** [review]

**Pull request merged**